common/transport/amqp/src/amqp_cbs.ts

```diff
@@ -0,0 +1,323 @@
+        attached: {
+          _onEnter: (callback) => {
+            if (callback) {
+              callback();
+            }
+            let tokenOperation = this._putTokenQueue.shift();
+            while (tokenOperation) {
+              this._fsm.handle('putToken', tokenOperation.audience, tokenOperation.token, tokenOperation.callback);
+              tokenOperation = this._putTokenQueue.shift();
+            }
+          },
+          attach: (callback) => callback(),
+          detach: (callback, err) => this._fsm.transition('detaching', callback, err),
+          forceDetach: () => {
+            this._receiverLink.forceDetach();
+            this._senderLink.forceDetach();
+            this._fsm.transition('detached');
+          },
+          putToken: (audience, token, putTokenCallback) => {
+          /*SRS_NODE_AMQP_CBS_16_011: [The `putToken` method shall construct an amqp message that contains the following application properties:
+          ```
+          'operation': 'put-token'
+          'type': 'servicebus.windows.net:sastoken'
+          'name': <audience>
+          ```
+          and system properties of
+          ```
+          'to': '$cbs'
+          'messageId': <uuid>
+          'reply_to': 'cbs'
+          ```
+          and a body containing `<sasToken>`.]*/
+            let amqpMessage = new AmqpMessage();
+            amqpMessage.applicationProperties = {
+              operation: 'put-token',
+              type: 'servicebus.windows.net:sastoken',
+              name: audience
+            };
+            amqpMessage.body = token;
+            amqpMessage.properties = {
+              to: '$cbs',
+              messageId: uuid.v4(),
+              reply_to: 'cbs'
+            };
+
+            let outstandingPutToken: PutTokenOperation = {
+              putTokenCallback: putTokenCallback,
+              expirationTime: Math.round(Date.now() / 1000) + this._putToken.numberOfSecondsToTimeout,
+              correlationId: amqpMessage.properties.messageId
+            };
+
+            this._putToken.outstandingPutTokens.push(outstandingPutToken);
+            //
+            // If this is the first put token then start trying to time it out.
+            //
+            if (this._putToken.outstandingPutTokens.length === 1) {
+              this._putToken.timeoutTimer = setTimeout(this._removeExpiredPutTokens.bind(this), this._putToken.putTokenTimeOutExaminationInterval);
+            }
+            /*Codes_SRS_NODE_AMQP_CBS_16_012: [The `putToken` method shall send this message over the `$cbs` sender link.]*/
+            this._senderLink.send(amqpMessage, (err) => {
+              if (err) {
+                // Find the operation in the outstanding array.  Remove it from the array since, well, it's not outstanding anymore.
+                // Since we may have arrived here asynchronously, we simply can't assume that it is the end of the array.  But,
+                // it's more likely near the end.
```

Comment:
>t's more likely near the end. [](start = 20, length = 29)

suggest you add a comment that the token might not even be in the array (if it expired).  That's not immediately obvious.